---
title: "Testing code differential expression"
output:
  html_notebook: default
  pdf_document: default
---
# Info 
executing a chunk  *Cmd+Shift+Enter*. 
Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.
When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

# Install libraries 
```{r}

# Tidyverse 
install.packages("tidyverse")

# DESeq
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq")

#ggplot2 
install.packages("ggplot2")

# Farver
install.packages("farver")

```

# Import libraries 
```{r}
library("DESeq")
library("ggplot2")
library("tidyverse")

```

# Load the data 
```{r}
# time = 0h
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_0h.txt',
                       sep = "\t", row.names = 1, header = TRUE)
colnames_0h_samples = str_remove(colnames(counts_0h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_0h_samples = paste("0h_",colnames_0h_samples,sep="")
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_0h.txt',
                       skip=5, row.names = 1) # skip first 5 lines
counts_0h = subset(counts_0h, select = c(Total, Total.1,Total.2))
colnames(counts_0h) = colnames_0h_samples

# time = 30min
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_30min.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_30min_samples = str_remove(colnames(counts_30min)[1:3],".sorted.bam") # keep the first 3 names 
colnames_30min_samples = paste("30min_",colnames_30min_samples,sep="")
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_30min.txt',
                          skip = 5, row.names = 1) # skip first 5 lines 
counts_30min = subset(counts_30min, select = c(Total, Total.1, Total.2))
colnames(counts_30min) = colnames_30min_samples
  
# time = 6h 
counts_6h =  read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_6h.txt',
                        sep = "\t", row.names = 1, header = TRUE)
colnames_6h_samples = str_remove(colnames(counts_6h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_6h_samples = str_remove(colnames_6h_samples,"mapped_")
colnames_6h_samples = paste("6h_",colnames_6h_samples,sep="")
counts_6h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_6h.txt',
                       skip = 5, row.names = 1) # skip first 5 lines
counts_6h = subset(counts_6h, select = c(Total, Total.1, Total.2))
colnames(counts_6h) = colnames_6h_samples

# time = 24h 
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_24h.txt',
                        sep = "\t", row.names = 1, header = TRUE)
colnames_24h_samples = str_remove(colnames(counts_24h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_24h_samples = str_remove(colnames_24h_samples,"mapped_")
colnames_24h_samples = paste("24h_",colnames_24h_samples,sep="")
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_24h.txt',
                        skip = 5, row.names = 1) # skip first 5 lines 
counts_24h = subset(counts_24h, select = c(Total, Total.1, Total.2))
colnames(counts_24h) = colnames_24h_samples

# remove colnames vectors 
rm(colnames_0h_samples,colnames_30min_samples,colnames_6h_samples,colnames_24h_samples)
```

# Amount of counts in total for each gene, remove later 
```{r}
counts_for_genes_0h = apply(counts_0h,1,sum)
counts_for_genes_30min = apply(counts_30min,1,sum)
counts_for_genes_6h = apply(counts_6h,1,sum)
counts_for_genes_24h = apply(counts_24h,1,sum)
summary(counts_for_genes_0h)
summary(counts_for_genes_30min)
summary(counts_for_genes_6h)
summary(counts_for_genes_24h)

# If we remove one gene for 0h + 6h, then the lines are not the same for 0h and 24h.. 
```

# DESEq, find differentially expressed genes 
```{r}

# TODO: is the filtering false default? 

# FDR cutoff 
cutoff = 0.05 

# time = 0h vs time = 30min (move above after)
counts_0h_30min_df = cbind(counts_0h,counts_30min)
times = factor(c("0h","0h","0h","30min","30min","30min")) # The two conditions
cds = newCountDataSet(counts_0h_30min_df,times) # Data structure for DESeq
cds = estimateSizeFactors(cds) # Normalization
cds = estimateDispersions(cds) # Estimate dispersion 
results = nbinomTest(cds, "0h","30min") # Calculate logfoldchange and p-values
#write.csv( results, file="differential_expression_results_0h_30min.csv" ) # Save output
significant_genes_0h_30min = results[results$padj < cutoff,] # Significant genes

# time = 0h vs time = 6h 
counts_0h_6h_df = cbind(counts_0h,counts_6h) # a dataframe with both timepoints
times = factor(c("0h","0h","0h","6h","6h","6h")) # The two conditions
cds = newCountDataSet(counts_0h_6h_df,times) # Data structure for DESeq
cds = estimateSizeFactors(cds) # Normalization
cds = estimateDispersions(cds) # Estimate dispersion 
results = nbinomTest(cds, "0h","6h") # Calculate logfoldchange and adj p-values
#write.csv( results, file="differential_expression_results_0h_6h.csv" ) # Save output
significant_genes_0h_6h = results[results$padj < cutoff,] # Signicicant genes 

# time = 0h vs time = 24h 
counts_0h_24h_df = cbind(counts_0h,counts_24h) # a dataframe with both timepoints
times = factor(c("0h","0h","0h","24h","24h","24h")) # The two conditions 
cds = newCountDataSet(counts_0h_24h_df,times) # Data structure for DESeq
cds = estimateSizeFactors(cds) # Normalization
cds = estimateDispersions(cds) # Estimate dispersion 
results = nbinomTest(cds,"0h","24h") # Calculate logfoldchange and adj p-values 
#write.csv(results, file = "differential_expression_results_0h_24h.csv") # Save output
significant_genes_0h_24h = results[results$padj < cutoff,] # Significant genes 

# Sort according to p-adj 
order_0h_30min = order(significant_genes_0h_30min$padj)
order_0h_6h = order(significant_genes_0h_6h$padj)
order_0h_24h = order(significant_genes_0h_24h$padj)

significant_genes_0h_30min = significant_genes_0h_30min[order_0h_30min,]
significant_genes_0h_6h = significant_genes_0h_6h[order_0h_6h,]
significant_genes_0h_24h = significant_genes_0h_24h[order_0h_24h,]

# Finding specific genes
#significant_genes_0h_30min[significant_genes_0h_30min$id=="sifB",]$log2FoldChange

```

# Testing something new DESEq, find differentially expressed genes 
# Normalize all counts in a common matrix, 
# get only 1 value of baseMean for time = 0h
```{r}

# TODO: is the filtering false default? 

# FDR cutoff 
cutoff = 0.05

counts_large_df = cbind(counts_0h,counts_30min,counts_6h,counts_24h)
times = factor(c("0h","0h","0h","30min","30min","30min","6h","6h","6h","24h","24h","24h")) # The conditions
cds = newCountDataSet(counts_large_df,times) # Data structure for DESeq
cds = estimateSizeFactors(cds) # Normalization
cds = estimateDispersions(cds) # Estimate dispersion

# Filtering (not used here)
#rs = rowSums(counts(cds))
#theta = 0.4
#use = (rs>quantile(rs,probs=theta))
#cdsFilt = cds[use,]
# use cdsFilt for nbinomTest()

# time = 0h vs time = 30min 
results = nbinomTest(cds, "0h","30min") # Calculate logfoldchange and p-values
#write.csv( results, file="differential_expression_results_0h_30min.csv" ) # Save output
significant_genes_0h_30min = results[results$padj < cutoff,] # Significant genes

# time = 0h vs time = 6h 
results = nbinomTest(cds, "0h","6h") # Calculate logfoldchange and adj p-values
#write.csv( results, file="differential_expression_results_0h_6h.csv" ) # Save output
significant_genes_0h_6h = results[results$padj < cutoff,] # Signicicant genes 

# time = 0h vs time = 24h 
results = nbinomTest(cds,"0h","24h") # Calculate logfoldchange and adj p-values 
#write.csv(results, file = "differential_expression_results_0h_24h.csv") # Save output
significant_genes_0h_24h = results[results$padj < cutoff,] # Significant genes 

# Sort according to p-adj 
order_0h_30min = order(significant_genes_0h_30min$padj)
order_0h_6h = order(significant_genes_0h_6h$padj)
order_0h_24h = order(significant_genes_0h_24h$padj)

significant_genes_0h_30min = significant_genes_0h_30min[order_0h_30min,]
significant_genes_0h_6h = significant_genes_0h_6h[order_0h_6h,]
significant_genes_0h_24h = significant_genes_0h_24h[order_0h_24h,]

# Finding specific genes
#significant_genes_0h_30min[significant_genes_0h_30min$id=="sifB",]$log2FoldChange

```


# Dot plot 
```{r}

# Larger than 1 or lower than -1 
high_0h_30min = sum(significant_genes_0h_30min$log2FoldChange > 1)
low_0h_30min = sum(significant_genes_0h_30min$log2FoldChange < (-1))

high_0h_6h = sum(significant_genes_0h_6h$log2FoldChange > 1)
low_0h_6h = sum(significant_genes_0h_6h$log2FoldChange < (-1))

high_0h_24h = sum(significant_genes_0h_24h$log2FoldChange > 1)
low_0h_24h = sum(significant_genes_0h_24h$log2FoldChange < (-1))

# Ratios reference 
ratio_ref_0h_30min = 8/28
ratio_ref_0h_6h = 271/155
ratio_ref_0h_24h = 148/112

# Ratios our 
ratio_our_0h_30min = high_0h_30min/low_0h_30min
ratio_our_0h_6h = high_0h_6h / low_0h_6h
ratio_our_0h_24h = high_0h_24h / low_0h_24h

# Dot plot 
logfold_30min = data.frame(group = "0.5",
                           value = significant_genes_0h_30min$log2FoldChange)
logfold_6h = data.frame(group="6", value = significant_genes_0h_6h$log2FoldChange)
logfold_24h = data.frame(group="24",value=significant_genes_0h_24h$log2FoldChange)
plot.data = rbind(logfold_30min, logfold_6h, logfold_24h)
plot.data$colors = 0
plot.data[which(plot.data$value < (-1)),"colors"] = -1
plot.data[which( plot.data$value > 1),"colors"] = 1
plot.data$colors = as.factor(plot.data$colors)

# Number of observations in each dot plot
n_fun <- function(x){
  return(data.frame(y = 6.5,
                    label = length(x)))
}

# Number of observations > 1 in each dot plot
n_fun_high <- function(x){
  return(data.frame(y = 1.5,
                    label = length(x[ x > 1])))
}

# Number of observations < (-1) in each dot plot
n_fun_low <- function(x){
  return(data.frame(y = -1.5,
                    label = length(x[ x <(-1)])))
}
# Plot it 
p = ggplot(plot.data, aes(x=group, y=value)) + 
  geom_dotplot(binaxis='y', stackdir='center',binwidth = 0.09,dotsize = 0.2, aes(color=colors)) +
  xlab("hours in water") + scale_y_continuous(name="log2 fold change in expression", breaks=seq(-5,6,1),limits=c(-5,7)) + 
  geom_hline(yintercept=1,linetype="dashed",size=0.5) + geom_hline(yintercept=-1,linetype="dashed",size=0.5) +
  theme_light() + scale_color_manual(values=c("blue", "black", "red")) + guides(color=FALSE) +
    stat_summary(fun.data = n_fun, geom = "text") + stat_summary(aes(x = as.numeric(group) + 0.2),fun.data = n_fun_high, geom = "text",color="red") +
  stat_summary(aes(x = as.numeric(group) + 0.2),fun.data = n_fun_low, geom = "text",color ="blue") +
  theme(axis.text.x=element_text(size=12,face="bold"))
p

```


# Heatmap 
```{r}
# Among significant genes keep those that have foldchange >2
# which is equivalent to logfoldchange > 1
sign_2_fold_geneID_0h_30min = significant_genes_0h_30min[significant_genes_0h_30min$foldChange > 2,]$id
sign_2_fold_geneID_0h_6h = significant_genes_0h_6h[significant_genes_0h_6h$foldChange > 2,]$id
sign_2_fold_geneID_0h_24h = significant_genes_0h_24h[significant_genes_0h_24h$foldChange > 2,]$id

# not used 
# < folchange 1/2 
# which is equivalent to logfoldchange < (-1)
#sign_2_fold_geneID_0h_30min = significant_genes_0h_30min[significant_genes_0h_30min$foldChange < (1/2),]$id
#sign_2_fold_geneID_0h_6h = significant_genes_0h_6h[significant_genes_0h_6h$foldChange < (1/2),]$id
#sign_2_fold_geneID_0h_24h = significant_genes_0h_24h[significant_genes_0h_24h$foldChange < (1/2),]$id

# Save unique geneIDs for all three comparisons
geneIDs_2_fold= c(sign_2_fold_geneID_0h_30min,sign_2_fold_geneID_0h_6h,
                         sign_2_fold_geneID_0h_24h)
unique_geneIDs_2_fold = unique(geneIDs_2_fold)

# Keep genes that are significant in all 3 comparisons 
intersect_0h_30min = intersect(significant_genes_0h_30min$id,unique_geneIDs_2_fold)
intersect_0h_30min_0h_6h = intersect(significant_genes_0h_6h$id,intersect_0h_30min)
all_sign_geneIDs_2_fold = intersect(significant_genes_0h_24h$id,intersect_0h_30min_0h_6h)

# Find basemean B for all these genes in all 3 comparisons, + basemean A for
# any of the 3 comparisons (if basemeanA is the same for all three comparisons)
sorted_all_sign_geneIDs_2_fold = sort(all_sign_geneIDs_2_fold)

# Order significant genes alphabetically 
alphabetically_significant_genes_0h_30min = significant_genes_0h_30min[order(significant_genes_0h_30min$id),]
alphabetically_significant_genes_0h_6h = significant_genes_0h_6h[order(significant_genes_0h_6h$id),]
alphabetically_significant_genes_0h_24h = significant_genes_0h_24h[order(significant_genes_0h_24h$id),]


clustering_0h_30min = alphabetically_significant_genes_0h_30min[sapply(alphabetically_significant_genes_0h_30min$id,is.element,sorted_all_sign_geneIDs_2_fold),]
clustering_0h = subset(clustering_0h_30min,select=c(id,baseMeanA))
clustering_30min = subset(clustering_0h_30min,select=c(id,baseMeanB))

clustering_0h_6h = alphabetically_significant_genes_0h_6h[sapply(alphabetically_significant_genes_0h_6h$id,is.element,sorted_all_sign_geneIDs_2_fold),]
clustering_6h = subset(clustering_0h_6h,select=c(id,baseMeanB))

clustering_0h_24h = alphabetically_significant_genes_0h_24h[sapply(alphabetically_significant_genes_0h_24h$id,is.element,sorted_all_sign_geneIDs_2_fold),]
clustering_24h = subset(clustering_0h_24h,select=c(id,baseMeanB))

# Create data frame 
basemean_df = cbind(clustering_0h$baseMeanA,clustering_30min$baseMeanB,clustering_6h$baseMeanB,clustering_24h$baseMeanB)
rownames(basemean_df) = clustering_0h$id
colnames(basemean_df) = c("0h","0.5h","6h","24h")
#basemean_df = log2(basemean_df) # Remove ?? 
# Create matrix 
basemean_matrix_2_fold = as.matrix(basemean_df)

# Heatmap 
# Eucledian is default distance matrix
heatmap(basemean_matrix_2_fold)


# TODO: Add colors; 
# High basemean = Red. Low basemean = Blue.


```

# Results deviates even more from original study, using RPKM values
# RPKM load data
```{r}
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_0h',
                       sep = "\t", row.names = 1, header = TRUE)
colnames_0h_samples = str_remove(colnames(counts_0h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_0h_samples = paste0("0h_",colnames_0h_samples)
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_0h',
                       skip=5, row.names = 1) # skip first 5 lines
counts_0h = subset(counts_0h, select = c(Total, Total.1,Total.2))
colnames(counts_0h) = colnames_0h_samples
counts_0h = round(counts_0h)

# time = 30min
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_values_30min.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_30min_samples = str_remove(colnames(counts_30min)[1:3],".sorted.bam") # keep the first 3 names
colnames_30min_samples = paste0("30min_",colnames_30min_samples)
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_values_30min.txt',
                          skip = 5, row.names = 1) # skip first 5 lines 
counts_30min = subset(counts_30min, select = c(Total, Total.1, Total.2))
colnames(counts_30min) = colnames_30min_samples
counts_30min = round(counts_30min)

# time = 6h
counts_6h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_6h.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_6h_samples = str_remove(colnames(counts_6h)[1:3],".sorted.bam")
colnames_6h_samples = str_remove(colnames_6h_samples,"mapped_")
colnames_6h_samples = paste0("6h_",colnames_6h_samples)
counts_6h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_6h.txt', 
                       skip = 5, row.names = 1) # skip first 5 lines
counts_6h = subset(counts_6h,select=c(Total, Total.1, Total.2))
colnames(counts_6h) = colnames_6h_samples
counts_6h = round(counts_6h)

# time = 24h
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_24h.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_24h_samples = str_remove(colnames(counts_24h)[1:3],".sorted.bam")
colnames_24h_samples = str_remove(colnames_24h_samples,"mapped_")
colnames_24h_samples = paste0("24h_",colnames_24h_samples)
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_24h.txt', 
                       skip = 5, row.names = 1) # skip first 5 lines
counts_24h = subset(counts_24h,select = c(Total,Total.1,Total.2))
colnames(counts_24h) = colnames_24h_samples
counts_24h = round(counts_24h)

```
