---
title: "Project code, finding differentially expressed genes with DESEq"
output: html_notebook
  pdf_document: default
---

# Install libraries 
```{r}
# Tidyverse 
#install.packages("tidyverse")
# DESeq
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq")
# ggplot2 
# install.packages("ggplot2")
# Farver
# install.packages("farver")
# VennDiagram package
# install.packages("VennDiagram")
```

# Import libraries needed
```{r}
library("DESeq")
library("ggplot2")
library("tidyverse")
library(VennDiagram) # to make venn diagrammes from DE genes
library(RColorBrewer) # coloring Venn diagrammes
library("gplots")
```

# Load the data 
Data from the mapping were loaded, then processed to remove the sense/antisense counts and only keep the total ones. 
```{r}
# time = 0h
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_0h.txt',
                       sep = "\t", row.names = 1, header = TRUE)
colnames_0h_samples = str_remove(colnames(counts_0h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_0h_samples = paste("0h_",colnames_0h_samples,sep="")
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_0h.txt',
                       skip=5, row.names = 1) # skip first 5 lines
counts_0h = subset(counts_0h, select = c(Total, Total.1,Total.2))
colnames(counts_0h) = colnames_0h_samples

# time = 30min
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_30min.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_30min_samples = str_remove(colnames(counts_30min)[1:3],".sorted.bam") # keep the first 3 names 
colnames_30min_samples = paste("30min_",colnames_30min_samples,sep="")
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_30min.txt',
                          skip = 5, row.names = 1) # skip first 5 lines 
counts_30min = subset(counts_30min, select = c(Total, Total.1, Total.2))
colnames(counts_30min) = colnames_30min_samples
  
# time = 6h 
counts_6h =  read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_6h.txt',
                        sep = "\t", row.names = 1, header = TRUE)
colnames_6h_samples = str_remove(colnames(counts_6h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_6h_samples = str_remove(colnames_6h_samples,"mapped_")
colnames_6h_samples = paste("6h_",colnames_6h_samples,sep="")
counts_6h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_6h.txt',
                       skip = 5, row.names = 1) # skip first 5 lines
counts_6h = subset(counts_6h, select = c(Total, Total.1, Total.2))
colnames(counts_6h) = colnames_6h_samples

# time = 24h 
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_24h.txt',
                        sep = "\t", row.names = 1, header = TRUE)
colnames_24h_samples = str_remove(colnames(counts_24h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_24h_samples = str_remove(colnames_24h_samples,"mapped_")
colnames_24h_samples = paste("24h_",colnames_24h_samples,sep="")
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/count_values/counts_values_24h.txt',
                        skip = 5, row.names = 1) # skip first 5 lines 
counts_24h = subset(counts_24h, select = c(Total, Total.1, Total.2))
colnames(counts_24h) = colnames_24h_samples
# remove colnames vectors 
rm(colnames_0h_samples,colnames_30min_samples,colnames_6h_samples,colnames_24h_samples)
```

# DESEq, find differentially expressed genes
Normalize all counts at the same time, in one dataframe 
```{r}

cutoff = 0.05 # FDR cutoff 
counts_large_df = cbind(counts_0h,counts_30min,counts_6h,counts_24h)
times = factor(c("0h","0h","0h","30min","30min","30min","6h","6h","6h","24h","24h","24h")) # The conditions
cds = newCountDataSet(counts_large_df,times) # Data structure for DESeq
cds = estimateSizeFactors(cds) # Normalization
cds = estimateDispersions(cds) # Estimate dispersion

# time = 0h vs time = 30min 
results_0h_30min = nbinomTest(cds, "0h","30min") # Calculate logfoldchange and p-values
significant_genes_0h_30min = results_0h_30min[results_0h_30min$padj < cutoff,] # Significant genes

# time = 0h vs time = 6h 
results_0h_6h = nbinomTest(cds, "0h","6h") # Calculate logfoldchange and adj p-values
significant_genes_0h_6h = results_0h_6h[results_0h_6h$padj < cutoff,] # Signicicant genes 

# time = 0h vs time = 24h 
results_0h_24h = nbinomTest(cds,"0h","24h") # Calculate logfoldchange and adj p-values 
significant_genes_0h_24h = results_0h_24h[results_0h_24h$padj < cutoff,] # Significant genes 

# Sort according to p-adj 
order_0h_30min = order(significant_genes_0h_30min$padj)
order_0h_6h = order(significant_genes_0h_6h$padj)
order_0h_24h = order(significant_genes_0h_24h$padj)

significant_genes_0h_30min = significant_genes_0h_30min[order_0h_30min,]
significant_genes_0h_6h =  significant_genes_0h_6h[order_0h_6h,]
significant_genes_0h_24h = significant_genes_0h_24h[order_0h_24h,]

# table with the significant gens and their output log2FCs and padjusted etc are saved as a text file. 
# Create new tables for the significant genes which have been processed together
write.table(significant_genes_0h_30min, file ="/home/student5/project/results/significant_genes_30min", sep = "\t",row.names=FALSE  )
write.table(significant_genes_0h_6h, file ="/home/student5/project/results/significant_genes_6h", sep = "\t",row.names=FALSE   )
write.table(significant_genes_0h_24h, file ="/home/student5/project/results/significant_genes_24h", sep = "\t",row.names=FALSE  )

```

# Create list of gene names and use in Venn diagramme
To access what genes were commonly differentially expressed at all time points, a venn diagramme was created. 
```{r}
# Create list of GeneIDs
genelist_30min = as.character(significant_genes_0h_30min$id)
genelist_6h = as.character(significant_genes_0h_6h$id)
genelist_24h = as.character(significant_genes_0h_24h$id)
myCol <- brewer.pal(3, "Pastel2")
# Chart where overlap between geneIDs are searched for
venn.diagram(
        x = list(genelist_30min, genelist_6h, genelist_24h),
        category.names = c("30 min" , "6h " , "24h"),
        filename = '/home/student5/project/results/figures/VennDiagram.png',
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)
```

# Dotplot
```{r}

# Dot plot 
logfold_30min = data.frame(group = "0.5",
                           value = significant_genes_0h_30min$log2FoldChange)
logfold_6h = data.frame(group="6", value = significant_genes_0h_6h$log2FoldChange)
logfold_24h = data.frame(group="24",value=significant_genes_0h_24h$log2FoldChange)
plot.data = rbind(logfold_30min, logfold_6h, logfold_24h)
plot.data$colors = 0
plot.data[which(plot.data$value < (-1)),"colors"] = -1
plot.data[which( plot.data$value > 1),"colors"] = 1
plot.data$colors = as.factor(plot.data$colors)

# Number of observations in each dot plot
n_fun <- function(x){
  return(data.frame(y = 6.5,
                    label = length(x)))
}
# Number of observations > 1 in each dot plot
n_fun_high <- function(x){
  return(data.frame(y = 1.5,
                    label = length(x[ x > 1])))
}
# Number of observations < (-1) in each dot plot
n_fun_low <- function(x){
  return(data.frame(y = -1.5,
                    label = length(x[ x <(-1)])))
}
# Plot it 
p = ggplot(plot.data, aes(x=group, y=value)) + 
  geom_dotplot(binaxis='y', stackdir='center',binwidth = 0.09,dotsize = 0.21, aes(color=colors)) +
  xlab("hours in water") + scale_y_continuous(name="log2 fold change in expression", breaks=seq(-5,6,1),limits=c(-5,8)) + 
  geom_hline(yintercept=1,linetype="dashed",size=0.5) + geom_hline(yintercept=-1,linetype="dashed",size=0.5) +
   scale_color_manual(values=c("blue", "black", "red")) + guides(color=FALSE) +
    stat_summary(fun.data = n_fun, geom = "text",size=8.5,position=position_nudge( y = 0.75)) + stat_summary(aes(x = as.numeric(group) + 0.4),fun.data = n_fun_high, geom = "text",color="red",size=8.5) +
  stat_summary(aes(x = as.numeric(group) + 0.4),fun.data = n_fun_low, geom = "text",color ="blue",size=8.5)  +theme_bw() + theme(axis.text.x=element_text(size=18),axis.text.y=element_text(size=18),panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +theme(axis.title.y = element_text(size = rel(1.8), angle = 90)) + theme(axis.title.x = element_text(size = rel(1.8), angle = 00))
 
pdf("/home/student5/project/results/figures/dotplot.pdf") 
p
dev.off()

```

# Heatmap 
Using all genes with both padj < 0.05 and foldchange > 2 for at least 1 timepoint (= 291 genes)
```{r}
# Among significant genes keep those that have foldchange >2
sign_2_fold_geneID_0h_30min = significant_genes_0h_30min[significant_genes_0h_30min$foldChange > 2,]$id
sign_2_fold_geneID_0h_6h = significant_genes_0h_6h[significant_genes_0h_6h$foldChange > 2,]$id
sign_2_fold_geneID_0h_24h = significant_genes_0h_24h[significant_genes_0h_24h$foldChange > 2,]$id

# Save unique geneIDs for all three comparisons
geneIDs_2_fold= c(sign_2_fold_geneID_0h_30min,sign_2_fold_geneID_0h_6h,
                         sign_2_fold_geneID_0h_24h)
unique_geneIDs_2_fold = unique(geneIDs_2_fold)

# Keep all genes in unique-vector
clustering_0h_30min = results_0h_30min[sapply(results_0h_30min$id,is.element,unique_geneIDs_2_fold),]
clustering_0h_6h = results_0h_6h[sapply(results_0h_6h$id,is.element,unique_geneIDs_2_fold),]
clustering_0h_24h = results_0h_24h[sapply(results_0h_24h$id,is.element,unique_geneIDs_2_fold),]

# Create matrix
basemean_matrix = cbind(clustering_0h_30min$baseMeanA,
                                     clustering_0h_30min$baseMeanB,
                                     clustering_0h_6h$baseMeanB,
                                     clustering_0h_24h$baseMeanB)

rownames(basemean_matrix) = clustering_0h_30min$id
colnames(basemean_matrix) = c("0h","0.5h","6h","24h")
basemean_matrix = log(basemean_matrix+1) # log transform

# Choose colors for heatmap 
redblue <- colorspace::diverge_hsv(100)
pal <- colorRampPalette(redblue)(100)

# Heatmap, Eucledian is default distance matrix
pdf("/home/student5/project/results/figures/heatmap.pdf",height=10,width=10) 
heatmap.2(basemean_matrix,col=pal,cexRow=0.01,density.info = "none",trace="none")
dev.off()
```


# RPKM load data 
Results deviates even more from original study, when using the RPKM values in DESeq like above
```{r}
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_0h',
                       sep = "\t", row.names = 1, header = TRUE)
colnames_0h_samples = str_remove(colnames(counts_0h)[1:3],".sorted.bam") # keep the first 3 names 
colnames_0h_samples = paste0("0h_",colnames_0h_samples)
counts_0h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_0h',
                       skip=5, row.names = 1) # skip first 5 lines
counts_0h = subset(counts_0h, select = c(Total, Total.1,Total.2))
colnames(counts_0h) = colnames_0h_samples
counts_0h = round(counts_0h)
# time = 30min
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_values_30min.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_30min_samples = str_remove(colnames(counts_30min)[1:3],".sorted.bam") # keep the first 3 names
colnames_30min_samples = paste0("30min_",colnames_30min_samples)
counts_30min = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_values_30min.txt',
                          skip = 5, row.names = 1) # skip first 5 lines 
counts_30min = subset(counts_30min, select = c(Total, Total.1, Total.2))
colnames(counts_30min) = colnames_30min_samples
counts_30min = round(counts_30min)
# time = 6h
counts_6h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_6h.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_6h_samples = str_remove(colnames(counts_6h)[1:3],".sorted.bam")
colnames_6h_samples = str_remove(colnames_6h_samples,"mapped_")
colnames_6h_samples = paste0("6h_",colnames_6h_samples)
counts_6h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_6h.txt', 
                       skip = 5, row.names = 1) # skip first 5 lines
counts_6h = subset(counts_6h,select=c(Total, Total.1, Total.2))
colnames(counts_6h) = colnames_6h_samples
counts_6h = round(counts_6h)
# time = 24h
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_24h.txt',
                          sep="\t", row.names = 1, header = TRUE)
colnames_24h_samples = str_remove(colnames(counts_24h)[1:3],".sorted.bam")
colnames_24h_samples = str_remove(colnames_24h_samples,"mapped_")
colnames_24h_samples = paste0("24h_",colnames_24h_samples)
counts_24h = read.table('/home/student5/project/mapping/mapping_2020_02_24/RPKM_values/RPKM_24h.txt', 
                       skip = 5, row.names = 1) # skip first 5 lines
counts_24h = subset(counts_24h,select = c(Total,Total.1,Total.2))
colnames(counts_24h) = colnames_24h_samples
counts_24h = round(counts_24h)
```
